internal/services/tesla_service.go

```diff
@@ -82,3 +86,101 @@ type TeslaVehicle struct {
 	VehicleID int    `json:"vehicle_id"`
 	VIN       string `json:"vin"`
 }
+
+type GetVehiclesResponse struct {
+	Response *[]TeslaVehicle `json:"response"`
+}
+
+type TeslaError struct {
+	Error            string
+	ErrorDescription string `json:"error_description"`
+	ReferenceID      string
+}
+
+var teslaScopes = []string{"user_data", "vehicle_device_data", "vehicle_cmds", "vehicle_charging_cmds", "energy_device_data", "energy_device_data", "energy_cmds"}
+
+type TeslaAuthCodeResponse struct {
+	AccessToken  string `json:"access_token"`
+	RefreshToken string `json:"refresh_token"`
+	IDToken      string `json:"id_token"`
+	ExpiresIn    int    `json:"expires_in"`
+	State        string `json:"state"`
+	TokenType    string `json:"token_type"`
+}
+
+func (t *teslaService) CompleteTeslaAuthCodeExchange(authCode, redirectURI, region string) (*TeslaAuthCodeResponse, error) {
+	u, _ := url.Parse("https://auth.tesla.com/oauth2/v3/token")
+	f := url.Values{}
+	f.Set("grant_type", "authorization_code")
+	f.Set("client_id", t.Settings.TeslaAuthorization.ClientID)
+	f.Set("client_secret", t.Settings.TeslaAuthorization.ClientSecret)
+	f.Set("code", authCode)
+	f.Set("redirect_uri", redirectURI)
+	f.Set("scope", "openid offline_access "+strings.Join(teslaScopes, " "))
+	f.Set("audience", fmt.Sprintf("https://fleet-api.prd.%s.vn.cloud.tesla.com", region))
+
+	req, err := http.NewRequest("POST", u.String(), strings.NewReader(f.Encode()))
```

Comment:
We May want to use a context here so this request doesn't hang our program.
Ideally, this function would take in a context, and we can use that directly.
we can also use the http.MethodPost constant.
```suggestion
	ctx, cancel := context.WithTimeout(context.Background(), time.Second*30) 
	defer cancel()
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, u.String(), strings.NewReader(f.Encode()))
```

Reply:
Definitely let's do this.